src/metabase/lib/js.cljs

```diff
@@ -427,6 +431,153 @@
+  (let [new-columns->columns
+        (zipmap (mapv (fn [field-ref]
+                        (if (and (vector? field-ref)
+                                 (= (first field-ref) :field)
+                                 (integer? (field-ref 2)))
+                          (dissoc (lib.equality/resolve-field-id a-query stage-number (field-ref 2))
+                                  :id)
+                          field-ref))
+                      columns)
+                columns)
+
+        matches
+        (lib.equality/find-closest-matches-for-refs
+         a-query
+         stage-number
+         field-refs
+         (mapv key new-columns->columns)
+         {:keep-join? true})]
+    (update-keys matches new-columns->columns)))
+
+(defn ^:export find-column-indexes-from-legacy-refs
+  "Given a list of columns (either JS `data.cols` or MLv2 `ColumnMetadata`) and a list of legacy refs, find each ref's
+  corresponding index into the list of columns.
+
+  Returns a parallel list to the refs, with the corresponding index, or -1 if no matching column is found."
+  [a-query stage-number legacy-columns legacy-refs]
+  ;; Set up this query stage's `:aggregation` list as the context for [[convert/->pMBQL]] to convert legacy
+  ;; `[:aggregation 0]` refs into pMBQL `[:aggregation uuid]` refs.
+  (convert/with-aggregation-list (:aggregation (lib.util/query-stage a-query stage-number))
+    (let [columns       (mapv ->column-or-ref legacy-columns)
+          field-refs    (map legacy-ref->pMBQL legacy-refs)
+          ;; matches is a map of columns to the corresponding index in field-refs.
+          matches       (match-many field-refs
+                                    columns
+                                    [;; first try matching by IDs keeping the join information
+                                     find-closest-matches-for-refs-by-id-and-join
+                                     ;; then try a normal match
+                                     #(lib.equality/find-closest-matches-for-refs a-query stage-number %1 %2 {:keep-join? true})
+                                     ;; then try matching field refs having a name instead of an ID
+                                     #(find-closest-matches-for-refs-by-name a-query stage-number %1 %2)])
+          ;; We want to return a parallel list to field-refs, giving the index of the matching column (or -1).
+          ;; First, map each column to its index (in the column list).
+          column->index (into {} (for [index (range (count columns))]
+                                   [(nth columns index) index]))
+          ;; And use that to map each match's ref-index to its column-index.
+          by-index      (into {} (for [[column ref-index] matches]
+                                   [ref-index (column->index column)]))
+          res (->> (range (count legacy-refs))
+                   (map #(by-index % -1))
+                   to-array)]
+      res)))
+
```

Comment:
I really don't like putting all this logic in `lib.js`, especially without writing any extra tests for it. It's a lot harder to test vs if it's in a `.cljc` file. Plus it's not available if other code needs to build on it in the rest of Metabase lib. 

Something like this should go in a regular `.cljc` namespace with actual tests, with wrappers around it only for doing conversion between JS and Cljs types

Reply:
I agree with Cam. Most of this logic looks like it should already have been handled by `lib.equality/find-closest-matches-for-refs`.

Reply:
I does remove part of it, but not the whole. If we want to incorporate something like this in `lib.equality/find-closest-matches-for-refs`, then we'll have to make it optional too and probably disable it at all other call sites. Also note that the composition logic implemented here is a bit different. Here, we make sure that a match early in the series of tests cannot be overridden by a later match. I think `lib.equality/find-closest-matches-for-refs` just assumes there are no duplicate matches. (In fact, I've switched the order of arguments in a `merge` in that function to keep earlier matches, so there might not be a real difference anymore.)